library/VTKExtensions/Rendering/vtkF3DRenderer.cxx

```diff
@@ -626,93 +649,138 @@ void vtkF3DRenderer::ConfigureHDRI()
 }
 
 //----------------------------------------------------------------------------
-void vtkF3DRenderer::ConfigureHDRITexture()
+void vtkF3DRenderer::ConfigureHDRIReader()
 {
-  // Read HDRI when needed
-  this->HasValidHDRIFile = false;
+  if (!this->HasValidHDRIReader && (this->HDRISkyboxVisible || this->GetUseImageBasedLighting()))
+  {
+    this->UseDefaultHDRI = false;
+    this->HDRIReader = nullptr;
+    if (!this->HDRIFile.empty())
+    {
+      if (!vtksys::SystemTools::FileExists(this->HDRIFile, true))
+      {
+        F3DLog::Print(
+          F3DLog::Severity::Warning, std::string("HDRI file does not exist ") + this->HDRIFile);
+      }
+      else
+      {
+        this->HDRIReader = vtkSmartPointer<vtkImageReader2>::Take(
+          vtkImageReader2Factory::CreateImageReader2(this->HDRIFile.c_str()));
+        if (this->HDRIReader)
+        {
+          this->HDRIReader->SetFileName(this->HDRIFile.c_str());
+        }
+        else
+        {
+          F3DLog::Print(F3DLog::Severity::Warning,
+            std::string("Cannot open HDRI file ") + this->HDRIFile +
+              std::string(". Using default HDRI"));
+        }
+      }
+    }
+
+    if (!this->HDRIReader)
+    {
+      // No valid HDRI file have been provided, read the default HDRI
+      // TODO add support for memory buffer in the vtkHDRReader in VTK
+      // https://github.com/f3d-app/f3d/issues/935
+      this->HDRIReader = vtkSmartPointer<vtkPNGReader>::New();
+      this->HDRIReader->SetMemoryBuffer(F3DDefaultHDRI);
+      this->HDRIReader->SetMemoryBufferLength(sizeof(F3DDefaultHDRI));
+      this->UseDefaultHDRI = true;
+    }
+    this->HasValidHDRIReader = true;
+  }
+  this->HDRIReaderConfigured = true;
+}
 
-  if (!this->HDRIFile.empty())
+//----------------------------------------------------------------------------
+void vtkF3DRenderer::ConfigureHDRIHash()
+{
+#if VTK_VERSION_NUMBER >= VTK_VERSION_CHECK(9, 2, 20221220)
+  if (!this->HasValidHDRIHash && this->GetUseImageBasedLighting() && this->HasValidHDRIReader)
   {
-    if (!vtksys::SystemTools::FileExists(this->HDRIFile, true))
+    if (this->UseDefaultHDRI)
     {
-      F3DLog::Print(
-        F3DLog::Severity::Warning, std::string("HDRI file does not exist ") + this->HDRIFile);
+      // Use a random HDRIHash, no need to generate one
+      this->HDRIHash = "fa96df14ed5268a6dbe81478abf8aac2";
```

Comment:
That's odd. Do you really need to set a hash here? What happens if `HDRIHash` is empty?

Reply:
then you'd need to be able to create an empty directory to store the cache, which is not supported.

Alternatively I could go through the HDRIHashInvalid code path in that case, but that makes it more complex. This is the simplest implementation.

Reply:
It means you are creating a cache directory for the default HDRI?

Reply:
yes of course, Id not need a random hash if not. As I said, I could consider that the default HDRI cannot have a hash and force IBL texture computation if you think this is a better approach.

Reply:
that should work indeed.